staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -124,41 +146,171 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions = append(
+			pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions,
+			topology,
+		)
+	}
 
 	return nil
 }
 
-// translateTopology converts existing zone labels or in-tree topology to CSI topology.
-// In-tree topology has precedence over zone labels.
-func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+// translateTopologyFromInTreeToCSI converts existing zone labels or in-tree topology to CSI topology.
+// In-tree topology has precedence over zone labels. When both in-tree topology and zone labels exist
+// for a particular CSI topology, in-tree topology will be used.
+func translateTopologyFromInTreeToCSI(pv *v1.PersistentVolume, csiTopologyKey string) error {
 	// If topology is already set, assume the content is accurate
-	if len(getTopologyZones(pv, topologyKey)) > 0 {
+	if len(getTopologyValues(pv, csiTopologyKey)) > 0 {
 		return nil
 	}
 
-	zones := getTopologyZones(pv, v1.LabelFailureDomainBetaZone)
+	zoneLabel, _ := getTopologyLabel(pv)
+
+	// migrate topology node affinity
+	zones := getTopologyValues(pv, zoneLabel)
 	if len(zones) > 0 {
-		return replaceTopology(pv, v1.LabelFailureDomainBetaZone, topologyKey)
+		return replaceTopology(pv, zoneLabel, csiTopologyKey)
 	}
 
-	if label, ok := pv.Labels[v1.LabelFailureDomainBetaZone]; ok {
+	// if nothing is in the NodeAffinity, try to fetch the topology from PV labels
+	label, ok := pv.Labels[zoneLabel]
+	if ok {
 		zones = strings.Split(label, labelMultiZoneDelimiter)
 		if len(zones) > 0 {
-			return addTopology(pv, topologyKey, zones)
+			return addTopology(pv, csiTopologyKey, zones)
+		}
+	}
+
+	return nil
+}
+
+// getTopologyLabel checks if the kubernetes topology label used in this
+// PV is GA and return the zone/region label used.
+// The version checking follows the following orders
+// 1. Check NodeAffinity
+//   1.1 Check if zoneGA exists, if yes return GA labels
+//   1.2 Check if zoneBeta exists, if yes return Beta labels
+//   1.3 Check if regionGA exists, if yes return GA labels
+//   1.4 Check if regionBeta exists, if yes return Beta labels
+// 2. Check PV labels
+//   2.1 Check if zoneGA exists, if yes return GA labels
+//   2.2 Check if zoneBeta exists, if yes return Beta labels
+//   2.3 Check if regionGA exists, if yes return GA labels
+//   2.4 Check if regionBeta exists, if yes return Beta labels
+func getTopologyLabel(pv *v1.PersistentVolume) (zoneLabel string, regionLabel string) {
+
+	if zoneGA := TopologyKeyExist(v1.LabelTopologyZone, pv.Spec.NodeAffinity); zoneGA {
+		return v1.LabelTopologyZone, v1.LabelTopologyRegion
+	}
+	if zoneBeta := TopologyKeyExist(v1.LabelFailureDomainBetaZone, pv.Spec.NodeAffinity); zoneBeta {
+		return v1.LabelFailureDomainBetaZone, v1.LabelFailureDomainBetaRegion
+	}
+	if regionGA := TopologyKeyExist(v1.LabelTopologyRegion, pv.Spec.NodeAffinity); regionGA {
```

Comment:
I was thinking we don't need to check for region at all. Because if zone is missing, then we're in trouble.

Reply:
Sounds good. I can remove the region related checks and only check zone